core/unit_test/TestMultiGPU.hpp

```diff
@@ -181,4 +181,54 @@ void test_scratch(TEST_EXECSPACE exec0, TEST_EXECSPACE exec1) {
   ASSERT_EQ(error0, 0);
   ASSERT_EQ(error1, 0);
 }
+
+// Test that stream synchronization behavior for various GPU APIs matches the
+// assumptions made in Kokkos for multi gpu support, namely, that any stream (no
+// matter which device it is created on) can be synced from any device.
+struct AccumulateTag {};
+struct CheckTag {};
+
+template <int N>
+struct StreamSyncFunctor {
+  int *value;
+  int *check;
+
+  KOKKOS_FUNCTION
+  void operator()(const AccumulateTag &, const int) const {
+    for (int i = 0; i < N; ++i) Kokkos::atomic_inc(value);
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(const CheckTag &, const int) const { check[0] = value[0]; }
+};
+
+void test_stream_sync(TEST_EXECSPACE exec0, TEST_EXECSPACE exec1,
+                      const std::function<void(TEST_EXECSPACE &)> &set_device) {
+  // Create data in host pinned space
+  Kokkos::View<int, Kokkos::SharedHostPinnedSpace> value("value"),
+      check("check");
+
+  constexpr int N        = 1000;
```

Comment:
hmm have you checked if 1000 atomic ops is enough to guarantee the host has dispatched the `set_device`? It sounds like something the gpu would just blast through (but I am not sure)

Reply:
I checked the value of `value` in between each call in this function, and it was not updated until after that fence, so we had successfully switched to dev1 without the kernel completing.

Reply:
Is that guaranteed to be true independent of what combination of gpu and host are used?
Could you add your check in some form to the automated test?

Reply:
I am not sure if there is any official documentation that explains *when* `Cuda` synchronizes the host pinned variable when it's written from device.

However, since `Kokkos` uses `cudaHostAlloc`, I guess we can safely follow the rules of *zero-copy* memory [from here](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/#zero-copy). So I would say that synchronization is needed before reading the variable on host.

However, since latency of the host-device transfer will highly depend on the available hardware, I am not sure we can come it with a rock solid test.

Reply:
AFAIK it should propagate straight through L2 and be picked up from there to be written to the host memory. But I agree that CUDA is not explicit about when this is to be expected to be completed.

Yeah, looks like we need a sync here.